examples/benchmark/benchmark_utils.py

```diff
@@ -0,0 +1,111 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+import time
+from dataclasses import dataclass
+
+import numpy as np
+import psutil
+import torch.cuda
+from pynvml import *
+
+monitor_gpu = False
+if torch.cuda.is_available():
+    monitor_gpu = True
+    nvmlInit()
+
+
+@dataclass
+class ExperimentResult:
+    def __init__(
+        self, training_time: int, num_samples: int, checkpoint_results: [] = None
+    ):
+        self.training_time = training_time
+        self.num_samples = num_samples
+        self.throughput = self.num_samples / self.training_time
+        self.resource_data = {}
+        self.avg_resource_data = {}
+        self.checkpoint_results = checkpoint_results
+
+
+class ResourceMonitor:
+    """
+    Monitors cpu and gpu usage+memory.
+    Set sleep_time_s carefully to avoid perf degradations.
+    """
+
+    def __init__(self, sleep_time_s=0.05, gpu_device=0, chunk_size=25_000):
+        self.monitor_thread = None
+        self.resource_data = {
+            "cpu_util": np.zeros(chunk_size),
+            "cpu_mem": np.zeros(chunk_size),
+            "gpu_util": np.zeros(chunk_size),
+            "gpu_mem": np.zeros(chunk_size),
+        }
+        self.stop_event = threading.Event()
+        self.sleep_time_s = sleep_time_s
+        self.gpu_device = gpu_device
+        self.chunk_size = chunk_size
+        self.cur_index = 0
+
+    def _check_and_expand(self):
+        if self.cur_index >= len(self.resource_data["cpu_util"]) - 1:
+            for key, arr in self.resource_data.items():
+                self.resource_data[key] = np.concatenate(
+                    (arr, np.zeros(self.chunk_size))
+                )
+
+    def _monitor(self):
+        while not self.stop_event.is_set():
+            self._check_and_expand()
+            self.resource_data["cpu_util"][self.cur_index] = psutil.cpu_percent()
+            self.resource_data["cpu_mem"][
+                self.cur_index
+            ] = psutil.virtual_memory().percent
+
+            if monitor_gpu:
+                gpu_info = nvmlDeviceGetUtilizationRates(
+                    nvmlDeviceGetHandleByIndex(self.gpu_device)
+                )
+                gpu_mem_info = nvmlDeviceGetMemoryInfo(
+                    nvmlDeviceGetHandleByIndex(self.gpu_device)
+                )
+                self.resource_data["gpu_util"][self.cur_index] = gpu_info.gpu
+                self.resource_data["gpu_mem"][self.cur_index] = (
+                    gpu_mem_info.used / gpu_mem_info.total * 100
+                )
+            self.cur_index += 1
+            time.sleep(self.sleep_time_s)
+
+    def start(self):
+        self.monitor_thread = threading.Thread(target=self._monitor)
+        self.monitor_thread.start()
+
+    def stop(self):
+        self.stop_event.set()
+
+        for key, arr in self.resource_data.items():
+            self.resource_data[key] = arr[: self.cur_index]
```

Comment:
Could we use `del arr[self.cur_index:]` here?

Reply:
Is this suggestion for readbility or is it a performance concern? 

`del` doesn't work against numpy arrays, but `numpy.delete` does. The approach above should be equivalent in performance to delete though since these are both ["no-copy" operations](https://towardsdatascience.com/the-performance-advantage-of-no-copy-dataframe-operations-7bf8c565c9a0#:~:text=When%20you%20slice%20a%20NumPy,and%20strides%20into%20that%20data.).

Reply:
It wasn't a performance concern, more a readability one. It took a couple of seconds to work out what was being done here, and I think a mutable delete is more readable here. What do you think?